asab/metrics/prometheus.py

```diff
@@ -40,64 +40,109 @@ def get_labels(tags):
+	unit = tags.get("unit")
+	if unit:
+		unit = validate_format(unit)
+	help = tags.get("help")
 	labels_str = get_labels(tags)
-
-	for v_name, value in metric.get("Values").items():
-		if validate_value is False:
+	if type == "counter":
+		values_items = values.items()
+	if type == "gauge":
+		values_items = metric.get("Values").items()
+	for v_name, value in values_items:
+		if validate_value(value) is False:
 			L.warning("Invalid OpenMetrics format. Value must be float or integer. {} omitted.".format(m_name))
 			continue
 		else:
-			name = "_".join([m_name, v_name])
-			name = validate_format(name)
-			# If a unit is specified it MUST be provided in a UNIT metadata line. In addition, an underscore and the unit MUST be the suffix of the MetricFamily name.
-			if tags.get("unit"):
-				unit = tags.get("unit")
-				unit = validate_format(unit)
-				name += "_{}".format(unit)
-				metric_lines.append("# TYPE {} {}".format(name, type))
-				metric_lines.append("# UNIT {} {}".format(name, unit))
-			else:
-				unit = None
-				metric_lines.append("# TYPE {} {}".format(name, type))
-				L.warning("Invalid OpenMetrics format. Please, add 'unit' in 'Tags'.")
-
-			if tags.get("help"):
-				metric_lines.append("# HELP {} {}".format(name, tags.get("help")))
-			else:
-				L.warning("Invalid OpenMetrics format. Please, add 'help' in 'Tags'.")
-
+			name = get_full_name(m_name, v_name, unit)
+			metric_lines.append(translate_metadata(name, type, unit, help))
 			if type == "counter":
-				metricpoint = "_total"
+				metric_lines.append(translate_counter(name, labels_str, value, created))
 			if type == "gauge":
-				metricpoint = ""
-			if labels_str:
-				metric_lines.append("{}{}{} {}".format(name, metricpoint, labels_str, value))
-			else:
-				metric_lines.append("{}{} {}".format(name, metricpoint, value))
+				metric_lines.append(translate_gauge(name, labels_str, value))
+
 	metric_text = '\n'.join(metric_lines)
 	return metric_text
 
 
-def to_openmetrics(metrics_service):
+def gauge_to_openmetrics(metrics_service):
 	lines = []
-	for mname, metrics in metrics_service.Metrics.items():
-		if isinstance(metrics, asab.metrics.metrics.Counter):
-			counter_text = metric_to_text(metrics.rest_get(), type="counter")
-			lines.append(counter_text)
-
+	for metrics in metrics_service.Metrics.values():
 		if isinstance(metrics, asab.metrics.metrics.Gauge):
-			gauge_text = metric_to_text(metrics.rest_get(), type="gauge")
+			gauge_text = metric_to_text(metrics.rest_get(), "gauge")
 			lines.append(gauge_text)
-
-	lines.append("# EOF\n")
 	text = '\n'.join(lines)
 	return text
 
 
+class PrometheusTarget(asab.ConfigObject):
+
+	def __init__(self, svc, config_section_name, config=None):
+		super().__init__(config_section_name, config)
+		self.countertext = ""
+		self.MetricsService = svc
+
+	async def process(self, now, mlist):
+		counter_lines = []
+		for metric, values in mlist:
+			if isinstance(metric, asab.metrics.metrics.Counter):
```

Comment:
Don't check the instance ... it is a bad practice in general.

Every metrics class should implement `metric.get_open_metric()` - it could return `None` to indicate that metics to be skipped from OpenMetrics b/c it is not yet implemented.